### ISSUE TITLE
Extract data from Python object generated with NCBI Entrez

### ISSUE THREAD

**Issue by mbonsma** [goyalsid/phageParser#47]
Current use of this project involves a pipeline where you download a bunch of stuff from NCBI using `acc2gb.py` and then run some other scripts to parse stuff. We've run into problems because of large file sizes, and so @JJ suggested combining some things into a single script and working directly with python objects fetched from NCBI.

The current pipeline looks something like this:
```
cat bac_accession_list.txt | python acc2gb.py yourname@institute.com > output.txt
python trimGenbankDNA.py
perl cas_in_gb_v2.pl
```
What this does:
- takes a list of bacteria and archaea accession numbers, pulls their entire genome and annotations from the NCBI database
- removes the actual DNA sequences which are unnecessary
- determines which CRISPR type(s) are present in each organism by looking for specific Cas genes.

We're looking for someone with fondness for Python who can look into the python objects fetched from NCBI: info on the `read` function [here](http://biopython.org/DIST/docs/api/Bio.Entrez-module.html#read). Ideally, we can extract the info we need right from the Python object instead of saving these monstrous files and reading them back in with other scripts.

This will have applications in other areas of this project, specifically Issue #27

**Comment by JJ** [comment]
In the short term, I'll wait for someone with chops in Python to do it
(I'll try it myself if nobody does). But in the long run, have you heard
about Taverna? http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3692062/ You can
create workflows in many different languages and use it to connect many
local and remote tools.

**Comment by mbonsma** [comment]
That is so cool! I'll definitely read up on Taverna.

**Comment by JJ** [comment]
This produces 
```
Fetching 3 entries from GenBank: NC_015138, NC_000853, NC_000854
Name NC_015138, 9491 features 
Location [4418502:4421010](+) Product ['CRISPR-associated protein Cas3']
Name NC_000853, 3844 features 
Name NC_000854, 3524 features 
Location [778744:780457](-) Product ['CRISPR-associated helicase Cas3']
```
A bit of massaging would be needed but I think it pretty much does what is needed, right?

**Comment by mbonsma** [comment]
It looks good to me (and the test worked for me as well). So just so I make sure I understand what's going on: are regular expressions too slow to realistically work, or is it having huge intermediate files that's a problem?

**Comment by JJ** [comment]
The problem is dealing with huge files using regular expressions. There are tools in BioPython to deal with it that do all the parsing for you, creating data structures where you only have to look in parts of the data structure you're interested, not through all the file (which is what "dumb" regexes do). 
This could work too with intermediate files using the SeqIO module (Perl or Python) on them, and then using regexes in the particular tag (or whichever they are called) you are interested.

**Comment by mbonsma** [comment]
Awesome. This has opened my eyes to many ways all of this could be improved (#48 for one). Thank you @JJ!

**Comment by JJ** [comment]
My pleasure!

**Comment by mbonsma** [comment]
Implemented `Bio.SeqIO.parse` in [`bac_info_parser.py`](https://github.com/goyalsid/phageParser/blob/master/bac_info_parser.py). Works like a charm, although there were a few cases where it raised an error when reading through a bunch of files downloaded from Genbank.

Also, it doesn't work on files that were processed using `trimGenbankDNA.py`, so I'll get rid of that function because it just complicates things. 

For info:
```
from Bio.SeqIO import parse
help(parse)
```

**Comment by JJ** [comment]
Great!

2015-07-20 20:51 GMT+02:00 mbonsma <notifications@github.com>:

> Closed #47 <https://github.com/goyalsid/phageParser/issues/47>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/goyalsid/phageParser/issues/47#event-360336213>.
>



-- 
JJ
